### PR TITLE
ENH: Enable LoadROId with Openslide

### PR DESCRIPTION
**Pull request by kenza-bouzid**
Refactor LoadRoiD/LoadPandaRoid transforms to work with OpenSlide and removed params redundancy with LoadingParams class.

### CODE REVIEW

File: hi-ml-cpath/src/health_cpath/datasets/panda_dataset.py [hunk 1]
```diff
@@ -52,89 +38,3 @@ def __init__(self,
         self.dataset_df[self.IMAGE_COLUMN] = "train_images/" + slide_ids + ".tiff"
         self.dataset_df[self.MASK_COLUMN] = "train_label_masks/" + slide_ids + "_mask.tiff"
         self.validate_columns()
-
-
-# MONAI's convention is that dictionary transforms have a 'd' suffix in the class name
-class ReadImaged(MapTransform):
```
Comment (kenza-bouzid) [inline hi-ml-cpath/src/health_cpath/datasets/panda_dataset.py:58]: this transform is not used anywhere in the code base

File: hi-ml-cpath/src/health_cpath/datasets/panda_dataset.py [hunk 2]
```diff
@@ -52,89 +38,3 @@ def __init__(self,
...
-        self.kwargs = kwargs
-
-    def __call__(self, data: Dict) -> Dict:
-        for key in self.keys:
-            if key in data or not self.allow_missing_keys:
-                data[key] = self.reader.read(data[key], **self.kwargs)
-        return data
-
-
-class LoadPandaROId(MapTransform):
```
Comment (kenza-bouzid) [inline hi-ml-cpath/src/health_cpath/datasets/panda_dataset.py:74]: Renamed to LoadMaskRoid so that we can use it for other cohorts without confusion

File: hi-ml-cpath/src/health_cpath/preprocessing/loading.py [hunk 3]
```diff
@@ -71,49 +138,222 @@ class LoadROId(MapTransform):
...
-    def _get_bounding_box(self, slide_obj: 'CuImage') -> Tuple[box_utils.Box, float]:
-        # Estimate bounding box at the lowest resolution (i.e. highest level)
-        highest_level = slide_obj.resolutions['level_count'] - 1
-        scale = slide_obj.resolutions['level_downsamples'][highest_level]
-        slide = load_slide_at_level(self.reader, slide_obj, level=highest_level)
+    def _load_slide_at_level(self, slide_obj: _OpenSlideOrCuImage, level: int) -> np.ndarray:
+        """Load full slide array at the given magnification level.
+
+        This is a manual workaround for a MONAI bug (https://github.com/Project-MONAI/MONAI/issues/3415)
+        fixed in a currently unreleased PR (https://github.com/Project-MONAI/MONAI/pull/3417).
```
Comment (kenza-bouzid) [inline hi-ml-cpath/src/health_cpath/preprocessing/loading.py:158]: @dccastro can we safely remove this workaround? I see that you were the one who opened the issue

File: hi-ml-cpath/src/health_cpath/preprocessing/loading.py [hunk 4]
```diff
@@ -71,49 +138,222 @@ class LoadROId(MapTransform):
...
         data[SlideKey.ORIGIN] = origin
         data[SlideKey.SCALE] = scale
-        data[SlideKey.FOREGROUND_THRESHOLD] = threshold
+        data[SlideKey.FOREGROUND_THRESHOLD] = self.foreground_threshold
 
         image_obj.close()
         return data
+
+
+class LoadMaskROId(MapTransform, BaseLoadROId, Generic[_OpenSlideOrCuImage]):
```
Comment (kenza-bouzid) [inline hi-ml-cpath/src/health_cpath/preprocessing/loading.py:201]: Please make sure I haven't accidentally messed up something while moving this transform to this file. original in panda_dataset.

File: hi-ml-cpath/src/health_cpath/scripts/mount_azure_dataset.py [hunk 5]
```diff
@@ -16,9 +16,9 @@
 from health_azure.utils import get_workspace  # noqa: E402
 
 
-def mount_dataset(dataset_id: str) -> Any:
+def mount_dataset(dataset_id: str, tmp_root: str = "/tmp/datasets/") -> Any:
```
Comment (kenza-bouzid) [inline hi-ml-cpath/src/health_cpath/scripts/mount_azure_dataset.py:19]: this script seems to be broken, to be fixed in separate PR
 ValueError: Unable to make dataset 'panda available for a local run because no AzureML workspace has been provided. Provide a workspace, or set a folder for local execution.to be fixed in a separate PR.

File: hi-ml-cpath/testhisto/testhisto/utils/test_viz_utils.py [hunk 6]
```diff
@@ -295,13 +294,3 @@ def test_location_selected_tiles(level: int) -> None:
     assert max(tile_xs) <= slide_image.shape[2] // factor
     assert min(tile_ys) >= 0
     assert max(tile_ys) <= slide_image.shape[1] // factor
-
-
-@pytest.mark.parametrize("backend", ["cuCIM", "OpenSlide"])
-@pytest.mark.parametrize("wsi_has_mask", [True, False])
-def test_load_image_dict(wsi_has_mask: bool, backend: str) -> None:
```
Comment (kenza-bouzid) [inline hi-ml-cpath/testhisto/testhisto/utils/test_viz_utils.py:302]: this test wasn't meaningful, we weren't loading a slide, test_load_slide should cover it netter

File: hi-ml-cpath/src/health_cpath/datamodules/base_module.py [hunk 7]
```diff
@@ -273,47 +272,29 @@ class SlidesDataModule(HistoDataModule[SlidesDataset]):
...
-                    level=self.level,
-                    backend=self.backend,
-                    **self.wsi_reader_args,
-                ),
+                self.loading_params.get_load_roid_transform(),
                 self.tiling_params.get_tiling_transform(bag_size=self.bag_sizes[stage], stage=stage),
-                # GridPatchd returns stacked tiles (bag_size, C, H, W), however we need to split them into separate
-                # tiles to be able to apply augmentations on each tile independently
-                SplitDimd(keys=SlideKey.IMAGE, dim=0, keepdim=False, list_output=True),
+                self.tiling_params.get_split_transform(),
```
Comment (kenza-bouzid) [inline hi-ml-cpath/src/health_cpath/datamodules/base_module.py:293]: moved to tiling_params to be able to test it. Given that args are hardcoded, we test that they are indeed what we expect, easier to catch if something goes wrong

**Review by maxilse — APPROVED** [review]

File: hi-ml-cpath/src/health_cpath/utils/output_utils.py [hunk 8]
```diff
@@ -254,24 +255,22 @@ def should_save_test_outputs(self, is_global_rank_zero: bool = True) -> bool:
 class DeepMILOutputsHandler:
     """Class that manages writing validation and test outputs for DeepMIL models."""
 
-    def __init__(self, outputs_root: Path, n_classes: int, tile_size: int, level: int,
+    def __init__(self, outputs_root: Path, n_classes: int, tile_size: int,
                  class_names: Optional[Sequence[str]], primary_val_metric: MetricsKey,
                  maximise: bool, val_plot_options: Collection[PlotOption],
-                 test_plot_options: Collection[PlotOption], wsi_has_mask: bool = True,
-                 backend: str = "cuCIM", val_set_is_dist: bool = True) -> None:
+                 test_plot_options: Collection[PlotOption], loading_params: LoadingParams = LoadingParams(),
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/utils/output_utils.py:261]: Avoid mutable default arguments. Long example why can be found here: https://towardsdatascience.com/python-pitfall-mutable-default-arguments-9385e8265422

File: hi-ml-cpath/testhisto/testhisto/preprocessing/test_loading.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+from health_cpath.scripts.mount_azure_dataset import mount_dataset
+from health_cpath.utils.naming import SlideKey
+from health_ml.utils.common_utils import is_gpu_available
+from testhiml.utils_testhiml import DEFAULT_WORKSPACE
+
+no_gpu = not is_gpu_available()
+
+
+@pytest.mark.parametrize("roi_type", [r for r in ROIType])
+def test_get_traansform_args(roi_type: str) -> None:
```
Comment (ant0nsc) [inline hi-ml-cpath/testhisto/testhisto/preprocessing/test_loading.py:23]:
```suggestion
def test_get_transform_args(roi_type: str) -> None:
```

**Review by ant0nsc — APPROVED** [review]
I can only comment on the high-level workings, and that looks good. Daniel is probably better placed to comment on the mechanics of the transforms. Left a suggestion to maybe restructure classes.

File: hi-ml-cpath/src/health_cpath/scripts/mount_azure_dataset.py (hunk 5 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/scripts/mount_azure_dataset.py]: Fixed
